File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    static int N,K;
    static int [][] dp;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());

        dp = new int[N+1][K+1];

        for(int i = 1; i <= N; ++i)
        {
            st = new StringTokenizer(br.readLine());
            int w = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());

            for(int j = 1; j <= K; ++j)
            {
                dp[i][j] = j < w ? dp[i-1][j] : Math.max(dp[i-1][j],v + dp[i-1][j-w]);
            }
        }
        System.out.print(dp[N][K]);
    }
}

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.kt
import java.io.BufferedReader
import java.io.InputStreamReader

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val (n, k) = br.readLine().split(" ").map { it.toInt() }
    val dp = Array(k + 1) { 0 }
    repeat(n) {
        val (w, v) = br.readLine().split(" ").map { it.toInt() }

        for (i in k downTo w) {
            dp[i] = maxOf(dp[i], dp[i - w] + v)
        }
    }

    print(dp[k])
}